src/main/java/org/folio/circulation/services/FeeFineService.java

```
@@ -1,58 +1,164 @@
-import org.folio.circulation.domain.representations.FeeFineActionStorageRepresentation;
+import org.folio.circulation.domain.ServicePoint;
+import org.folio.circulation.domain.ServicePointRepository;
+import org.folio.circulation.domain.User;
+import org.folio.circulation.domain.UserRepository;
+import org.folio.circulation.domain.representations.StoredAccount;
+import org.folio.circulation.domain.representations.StoredFeeFineAction;
+import org.folio.circulation.services.feefine.AccountRefundContext;
+import org.folio.circulation.services.feefine.AccountRefundProcessor;
+import org.folio.circulation.services.support.CreateAccountCommand;
+import org.folio.circulation.services.support.RefundAccountCommand;
 import org.folio.circulation.support.Clients;
 import org.folio.circulation.support.Result;
 import org.folio.circulation.support.results.CommonFailures;
 
+import io.vertx.core.json.JsonObject;
+
 public class FeeFineService {
   private final AccountRepository accountRepository;
   private final FeeFineActionRepository feeFineActionRepository;
+  private final UserRepository userRepository;
+  private final ServicePointRepository servicePointRepository;
 
   public FeeFineService(Clients clients) {
     this.accountRepository = new AccountRepository(clients);
     this.feeFineActionRepository = new FeeFineActionRepository(clients);
+    this.userRepository = new UserRepository(clients);
+    this.servicePointRepository = new ServicePointRepository(clients);
   }
 
   public CompletableFuture<Result<Void>> createAccountsAndActions(
-    Collection<FeeFineAccountAndAction> accountAndActions) {
+    Collection<CreateAccountCommand> accountAndActions) {
 
     return allOf(accountAndActions.stream()
       .map(this::createAccountAndAction)
       .toArray(CompletableFuture[]::new))
-      .thenApply(notUsed -> Result.<Void>succeeded(null))
+      .thenApply(Result::succeeded)
       .exceptionally(CommonFailures::failedDueToServerError);
```

Why are these steps needed when the `createAccountAndAction` method returns a `CompletableFuture<Result<Void>>` already? Is this to convert a stream of results to a single result for all of them?

---

Yes, correct. The method returns single completable future which is successful only when all of the futures are completed successfully.

---

What happens during this process when one of the commands fails? From what I can tell, this will create a successful result, masking that failure.

---

`thenApply(Result::succeeded)` won't be executed when a command failed, but the `exceptionally(CommonFailures::failedDueToServerError)` method will be executed instead.

---

Does that mean that `createAccount` will throw an exception when it fails, rather than return a failed result?

As I understand it, the `thenApply` will only be skipped and `exceptionally` executed when an unhandled exception is thrown. Have I misunderstood?

---

Thanks for pointing on this. It was not working as I expected to. I've added unit test for both methods and fixed methods.